Mathlib/RingTheory/Ideal/Over.lean

```diff
@@ -526,21 +556,43 @@ theorem algebraMap_injective_of_liesOver :
   exact Quotient.eq.mp hab
 
 instance [P.IsPrime] : NoZeroSMulDivisors (A ⧸ p) (B ⧸ P) :=
-  NoZeroSMulDivisors.of_algebraMap_injective (Quotient.algebraMap_injective_of_liesOver P p)
-
-/-- If `P` lies over `p`, then the stabilizer of `P` acts on the extension `(B ⧸ P) / (A ⧸ p). -/
+  NoZeroSMulDivisors.of_algebraMap_injective (algebraMap_injective_of_liesOver P p)
+
+variable {p} in
+theorem nontrivial_of_liesOver_ne_top (hp : p ≠ ⊤) : Nontrivial (B ⧸ P) :=
+  Quotient.nontrivial ((eq_top_iff_liesOver_eq_top P p).mp.mt hp)
+
+theorem nontrivial_of_liesOver_isPrime [hp : p.IsPrime] : Nontrivial (B ⧸ P) :=
+  nontrivial_of_liesOver_ne_top P hp.ne_top
+
+variable {P} in
+/-- An `A ⧸ p`-algebra isomorphism between `B ⧸ P` and `C ⧸ Q` induced by an `A`-algebra
+  isomorphism between `B` and `C`, where `Q = σ P`. -/
+def algEquivOfEqMap {E : Type*} [EquivLike E B C] [AlgEquivClass E A B C] (σ : E)
+    (h : Q = P.map σ) : (B ⧸ P) ≃ₐ[A ⧸ p] (C ⧸ Q) := {
+  quotientEquiv P Q σ h with
+  commutes' := by
+    rintro ⟨x⟩
+    exact congrArg (Ideal.Quotient.mk Q) (AlgHomClass.commutes σ x)
+}
+
+variable {P} in
+/-- An `A ⧸ p`-algebra isomorphism between `B ⧸ P` and `C ⧸ Q` induced by an `A`-algebra
+  isomorphism between `B` and `C`, where `P = σ⁻¹ Q`. -/
+def algEquivOfEqComap {E : Type*} [EquivLike E B C] [AlgEquivClass E A B C] (σ : E)
+    (h : P = Q.comap σ) : (B ⧸ P) ≃ₐ[A ⧸ p] (C ⧸ Q) := by
+  rw [← show _ = Q.comap σ from map_symm (σ : B ≃+* C)] at h
+  exact (algEquivOfEqMap p (σ : B ≃ₐ[A] C).symm h).symm
```

Comment:
Using rewrite to construct data is very bad. It gives `Eq.cast` in the term that is very unusable. Consider using `Ideal.quotientEquivAlgOfEq`.